tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/BlobUploadProcessor.cs

```diff
@@ -887,6 +889,63 @@ private async Task UploadTestRunBlobAsync(string account, Build build, TestRun t
             }
         }
 
+        private async Task UploadTestRunResultsAsync(string account, Build build, TestRun testRun)
+        {
+            try
+            {
+                var continuationToken = string.Empty;
+                var buildIds = new[] { build.Id };
+                var discoveredTestResults = new List<ShallowTestCaseResult>();
+                var blobPath = $"{build.Project.Name}/{testRun.CompletedDate:yyyy/MM/dd}/{testRun.Id}/results.jsonl";
+                var blobClient = this.testRunsContainerClient.GetBlobClient(blobPath);
+
+                if (await blobClient.ExistsAsync())
+                {
+                    this.logger.LogInformation("Skipping existing test results blob for build {BuildId}, test run {RunId}", build.Id, testRun.Id);
+                    return;
+                }
+
+                do
+                {
+                    // the unfortunate fact of the matter is that testResultsClient does not have a API Surface that can get us
+                    // all test results for a specific TestRunId. As a result, we do some internal filtering to only grab the test results that
+                    // are available for a specific run. This will have the knock-on affect of meaning we hit that heavyweight API for test results
+                    // multiple times on each BuildCompletionEvent. This could be a problem, as I believe we listen to individual build completion, and not
+                    // pipeline completion events.
+                    // First build completes, only needs to pull test results for that build...
+                    // but as time goes on, and more builds complete, the set of test results that will be returned will be heavier and heavier.
+                    // we either implement our own manual REST request to devops, or we use metrics only. Leaving that aside for now, I know
+                    // which devops API I can go to manually, it just goes against the grain of the rest of this project.
```

Comment:
If @hallipr looks at this, this is large discovered concern. The API Surface is not great unfortunately.